### PR TITLE
Limit target amount

### PR DESCRIPTION
**Pull request by flomosq**
# Limit target amount per user

### Trello Card #5

- Validate max target amount per user
- Max target amount per user test

### CODE REVIEW

File: app/models/target.rb [hunk 1]
```diff
@@ -31,4 +33,15 @@ class Target < ApplicationRecord
...
+
+  validate :target_limit_per_user, on: :create
+
+  private
+
+  def target_limit_per_user
+    return if user.targets.count < MAX_TARGET_AMOUNT_PER_USER
+
+    errors.add(:targets,
+               "Target per user limit exceeded. Max allowed is #{MAX_TARGET_AMOUNT_PER_USER}")
```
Comment (santir489) [inline app/models/target.rb]: We should use I18n here 🙌

File: spec/models/target_spec.rb [hunk 2]
```diff
@@ -48,4 +48,16 @@
...
   end
+
+  context 'when the user has 10 targets' do
+    let(:user) { create(:user) }
+
+    before { create_list(:target, 10, user: user) }
+
+    subject { build(:target, user: user) }
+
+    it 'target is invalid' do
```
Comment (santir489) [inline spec/models/target_spec.rb]: maybe check the error message as well. WDYT?

File: app/models/target.rb [hunk 3]
```diff
@@ -31,4 +33,15 @@ class Target < ApplicationRecord
                        numericality: { greater_than_or_equal_to: -90, less_than_or_equal_to: 90 }
   validates :longitude, presence: true,
                         numericality: { greater_than_or_equal_to: -180, less_than_or_equal_to: 180 }
+
+  validate :target_limit_per_user, on: :create
```
Comment (santir489) [inline app/models/target.rb]: Maybe I am being too prickly, but... why only `on create`?

Comment (brunvez) [reply app/models/target.rb]: Yeah this should be just a normal validation. Speaking of validations, I'm going to be a prick too and say that I don't like method validations. I always prefer [Validators](https://guides.rubyonrails.org/active_record_validations.html#custom-validators), even for trivial cases like these.

File: config/locales/en.yml [hunk 4]
```diff
@@ -36,3 +36,7 @@ en:
       invalid_content_type: Invalid content type header
       missing_param: A required param is missing
       not_found: Couldn't find the record
+  models:
+    target:
+      errors:
+        targets_per_user_limit: Target per user limit exceeded. Max allowed is %{limit}
```
Comment (santir489) [inline config/locales/en.yml]: missing new end line here

File: app/models/validators/target_limit_per_user_validator.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+module Validators
+  class TargetLimitPerUserValidator < ActiveModel::Validator
```
Comment (santir489) [inline app/models/validators/target_limit_per_user_validator.rb:2]: 🚀

File: .reek.yml [hunk 6]
```diff
@@ -19,7 +19,7 @@ detectors:
   DuplicateMethodCall:
     enabled: true
     exclude: []
-    max_calls: 1
+    max_calls: 3
```
Comment (santir489) [inline .reek.yml:22]: mmm maybe we do not want to change this, we can explore another alternatives

File: app/models/validators/target_limit_per_user_validator.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    MAX_TARGET_AMOUNT_PER_USER = 10
+
+    def validate(record)
+      return unless record.user
+      return if record.user.targets.count < MAX_TARGET_AMOUNT_PER_USER
```
Comment (santir489) [inline app/models/validators/target_limit_per_user_validator.rb:7]: We can rewrite this.
Two return statements in one method, kind of smells...

File: app/models/target.rb [hunk 8]
```diff
@@ -23,6 +23,10 @@
 #  fk_rails_...  (user_id => users.id)
 #
 class Target < ApplicationRecord
+  include ActiveModel::Validations
+
+  validates_with Validators::TargetLimitPerUserValidator
```
Comment (afmicc) [inline app/models/target.rb]: take a look at our tech guides about the [order of different methods](https://github.com/rootstrap/tech-guides/blob/master/ruby/rails.md#macro-style-methods) inside classes

File: spec/models/target_spec.rb [hunk 9]
```diff
@@ -48,4 +48,24 @@
...
```
Comment (afmicc) [inline spec/models/target_spec.rb]:
```suggestion
    it 'is invalid' do
```

**Review by brunvez — APPROVED** [review]

**Review by afmicc — APPROVED** [review]

**Review by santir489 — APPROVED** [review]

**Pull request merged**